### PR TITLE
Eliminate data race in codec

### LINKED ISSUE #56: NewCodec triggers the race detector when run on different goroutines
Run the following test with -race on:

```go
func TestRaceCodecConstruction(t *testing.T) {

	comms := make(chan []byte, 1000)
	done := make(chan error, 10)

	go func() {
		recordSchemaJSON := `{"type": "long"}`
		codec, _ := NewCodec(recordSchemaJSON)
		for i := 0; i < 10000; i++ {

			bb := new(bytes.Buffer)
			if err := codec.Encode(bb, int64(i)); err != nil {
				done <- err
				return
			}

			comms <- bb.Bytes()
		}

		close(comms)
	}()

	go func() {
		recordSchemaJSON := `{"type": "long"}`
		codec, _ := NewCodec(recordSchemaJSON)

		i := 0
		for encoded := range comms {
			bb := bytes.NewBuffer(encoded)
			decoded, err := codec.Decode(bb)
			if err != nil {
				done <- err
				return
			}
			result := decoded.(int64)
			if result != int64(i) {
				done <- fmt.Errorf("didnt match %v %v", i, result)
				return
			}

			i++
		}

		close(done)
	}()

	err := <-done
	if err != nil {
		t.Fatal(err)
	}
}
```

This will trigger a race detected.:

```
==================
WARNING: DATA RACE
Write by goroutine 51:
  github.com/linkedin/goavro.NewCodec()
      ~/go/src/github.com/linkedin/goavro/codec.go:195 +0x5f9
  github.com/linkedin/goavro.TestRaceCodecConstruction.func1()
      ~/go/src/github.com/linkedin/goavro/race_test.go:16 +0x6c

Previous write by goroutine 52:
  github.com/linkedin/goavro.NewCodec()
      ~/go/src/github.com/linkedin/goavro/codec.go:195 +0x5f9
  github.com/linkedin/goavro.TestRaceCodecConstruction.func2()
      ~/go/src/github.com/linkedin/goavro/race_test.go:33 +0x69

Goroutine 51 (running) created at:
  github.com/linkedin/goavro.TestRaceCodecConstruction()
      ~/go/src/github.com/linkedin/goavro/race_test.go:29 +0xa4
  testing.tRunner()
      ~/opt/go/src/testing/testing.go:473 +0xdc

Goroutine 52 (running) created at:
  github.com/linkedin/goavro.TestRaceCodecConstruction()
      ~/go/src/github.com/linkedin/goavro/race_test.go:53 +0xd0
  testing.tRunner()
      ~/opt/go/src/testing/testing.go:473 +0xdc
==================
```

Nothing jumps out at me as being particularly troublesome in that code for concurrent use but the init section with static codecs is potentially an issue.

In practice I've also been able to get the race detector to trigger when in 1 goroutine codecs were being created and in another another codec was being used for decoding records.  It pointed at line 595 of the codec.go file.

### PR DESCRIPTION
**Pull request by kklipsch**
Proposed change to fix #56

### CODE REVIEW

File: codec.go [hunk 1]
```diff
@@ -124,7 +124,42 @@ func (c codec) String() string {
...
+		longCodec:    longCodec(),
+		floatCodec:   &codec{nm: &name{n: "float32"}, df: floatDecoder, ef: floatEncoder},
+		doubleCodec:  &codec{nm: &name{n: "float64"}, df: doubleDecoder, ef: doubleEncoder},
+		bytesCodec:   &codec{nm: &name{n: "[]uint8"}, df: bytesDecoder, ef: bytesEncoder},
+		stringCodec:  &codec{nm: &name{n: "string"}, df: stringDecoder, ef: stringEncoder},
+	}
+
+}
+
+func longCodec() *codec {
```
Comment (thomaso-mirodin) [inline codec.go]: Mind if I ask why this in a separate function?

Comment (kklipsch) [inline codec.go]: Its used in the reader/writer code as well.

Comment (thomaso-mirodin) [inline codec.go]: Nm! Makes sense :D

**Comment by thomaso-mirodin** [comment]
Thanks, lgtm from me

@karrick for a second review

**Comment by kklipsch** [comment]
One thing I'll note, this obviously adds allocation pressure and I didn't measure the delta.

**Comment by karrick** [comment]
Wow! It took me a while to see what was going on here. Funny part is that a long time ago I was considering doing nearly exactly this for other reasons while improving namespace handling, so you totally did a chunk of work for that effort.

Thanks a bunch!

**Comment by karrick** [comment]
The allocation pressure of a new codec is minuscule to the allocation of reading a single large datum from a stream.

But that's my project today. Flat-buffer style cursor providing you read-only access to a datum.

-Karrick

**Pull request merged**